### DISCUSSION TITLE
Wh/km not showing in 'Drives'

### DISCUSSION THREAD

**Discussion by morten-holm** [teslamate-org/teslamate#1411] [Q&A] [answered]
When navigating to teslamate -> Dashboards -> Drives, Wh/km is not always showing. Wh/km is showing in the details of that  drive.

It is not only short trips that does not show Wh/km in the list. In the last 7 days I have 36 drives and 10 of these has no Wh/km.

**Screenshots**

In this particular example it is drive nr. 2 on the list.

<img width="1421" alt="Screenshot 2021-02-22 at 20 12 09" src="https://user-images.githubusercontent.com/2429501/108758567-8c31c100-754b-11eb-8b36-36b95f3be2e7.png">
<img width="1422" alt="Screenshot 2021-02-22 at 20 12 28" src="https://user-images.githubusercontent.com/2429501/108758592-93f16580-754b-11eb-82ab-38423e1927d9.png">

Hope there is a quick fix. Please let me know if any data from pgsql is needed.

**Comment by joshentu** [comment]
Same issue here, with latest v1.21.5 installed

**Comment by DrMichael** [comment]
Well, there are some sanity checks, e.g. when distance < 1 or duration < 1, then nothing is shown here. That is obviously not the case here.

Could you please insert the attached query into the panel and post the result?

[drives.txt](https://github.com/adriankumpf/teslamate/files/6028381/drives.txt)

**Comment by morten-holm** [comment]
Like this?

<img width="1403" alt="Screenshot 2021-02-23 at 19 15 58" src="https://user-images.githubusercontent.com/2429501/108888352-a337e800-760b-11eb-8a8d-accd2e2bdf5c.png">

**Comment by DrMichael** [comment]
Yes, not clear to me, why `end_position.battery_level != end_position.usable_battery_level` leads to not showing the consumption value...

Wanted to redo the Dashboard with the new table anyway. I left that out, some values are now, well, high... :-)

Give it a try.
[Drives7.zip](https://github.com/adriankumpf/teslamate/files/6044507/Drives7.zip)

**Comment by BabyYeggie** [comment, marked as answer]
Consumption was set to null based on #257 where useable battery level cannot be reliably determined by the API. If you drive with a cold battery, it will eventually heat up and unlock useable range leading to unusually high efficiency/low consumption. I think my record was 1200% efficiency due to this effect.

**Comment by DrMichael** [comment]
But that does not really explain, why it is not shown as in the original post...

**Comment by BabyYeggie** [comment]
The null consumption and efficiency is from the check to see if the battery is cold or not. It's from this part of the table:
`case when (start_position.battery_level != start_position.usable_battery_level OR end_position.battery_level != end_position.usable_battery_level) = true then true else false end  as reduced_range,
    duration_min > 1 AND distance > 1 AND ( 
      start_position.usable_battery_level IS NULL OR end_position.usable_battery_level IS NULL	OR
      (end_position.battery_level - end_position.usable_battery_level) = 0 
    ) as is_sufficiently_precise`

In order for the consumption/efficiency to be displayed, column 1 or 3 needs to be true. If either is false, the null is displayed. In the current revision, the value of column 2 doesn't matter. (Referring to chart on post 4).

The "is_sufficiently_precise" is used to determine whether to show the results of the calculations or null.
` CASE WHEN is_sufficiently_precise THEN range_diff * car_efficiency / distance * 1000 * CASE WHEN '$length_unit' = 'km' THEN 1
                                                                                                WHEN '$length_unit' = 'mi' THEN 1.60934
                                                                                           END
    END AS consumption_kWh_$length_unit,
    CASE WHEN is_sufficiently_precise THEN distance / range_diff
         ELSE NULL
    END AS efficiency`


What we need is a "Drives" table for warm weather folks and another without the battery level check / "is_sufficiently_precise" for us cold weather people.

**Comment by DrMichael** [comment]
> The null consumption and efficiency is from the check to see if the battery is cold or not. It's from this part of the table:
> `case when (start_position.battery_level != start_position.usable_battery_level OR end_position.battery_level != end_position.usable_battery_level) = true then true else false end as reduced_range, duration_min > 1 AND distance > 1 AND ( start_position.usable_battery_level IS NULL OR end_position.usable_battery_level IS NULL OR (end_position.battery_level - end_position.usable_battery_level) = 0 ) as is_sufficiently_precise`

May I propose to change  `case when (start_position.battery_level != start_position.usable_battery_level OR end_position.battery_level != end_position.usable_battery_level) = true then true else false end as reduced_range` to just `(start_position.battery_level != start_position.usable_battery_level OR end_position.battery_level != end_position.usable_battery_level) as reduced_range`

> In order for the consumption/efficiency to be displayed, column 1 or 3 needs to be true. If either is false, the null is displayed. In the current revision, the value of column 2 doesn't matter. (Referring to chart on post 4).

But that means, that either one of `start_position.usable_battery_level` and `end_position.usable_battery_level` needs to be NULL or `end_position.battery_level` equal to `end_position.usable_battery_level`. The latter one appears far to sharp, a small difference should also be fine.

> The "is_sufficiently_precise" is used to determine whether to show the results of the calculations or null.
> ` CASE WHEN is_sufficiently_precise THEN range_diff * car_efficiency / distance * 1000 * CASE WHEN '$length_unit' = 'km' THEN 1 WHEN '$length_unit' = 'mi' THEN 1.60934 END END AS consumption_kWh_$length_unit, CASE WHEN is_sufficiently_precise THEN distance / range_diff ELSE NULL END AS efficiency`

That should probably also use `convert_km`...

**Comment by pkuehnel** [comment]
I don't get it. Why should consumption not be calculated correctly if `usable_battery_level` and `battery_level` are not the same? The real consumption can not be calculated with `usable_battery_level`, because it's dependent on temperature, but the `battery_level` is independent from temperature, so even if range is reduced due to cold temperature `battery_level` could be used for calculating consumption.

Or am I missing something? I thougt that comumed Energy would come from `start_position.battery_level`- `end_position.battery_level` * battery capacity

**Comment by DrMichael** [comment]
Yes, sorry, I also don't understand that...

@adriankumpf can you bring light into this?

**Reply by adriankumpf** [reply]
- `usable_battery_level` != `battery_level` is an indicator for a cold battery
- the consumption estimate is based on the range values, which are dependent on temperature
- if you drive with a cold battery, it warms up and thus increases the range, which in turn has the effect of distorting the range calculation.

**Reply by pkuehnel** [reply]
probably possible to use `battery_level` as fallback?

**Reply by pkuehnel** [reply]
well now I understand, the issue is, that you don't know the Degradation of the battery when using `battery_level` only. so you would have to save the calculated battery capacity anywhere because you can only calculate the capacity when the battery is not cold...
